Fixing  WideAngleCamera Lensflare issue

sanjuksha opened the pull request:
Signed-off-by: Sanjuksha Nirgude <sanjuksha@gmail.com>


# 🦟 Bug fix
https://github.com/gazebosim/gazebo-classic/issues/3273

## Summary
Porting the same change for gazebo classic from https://github.com/gazebosim/gz-rendering/pull/746  and a fix for wide angle camera not showing lensflare at certain angles. Wide Angle camera used to loop through all cameras to find the occlusion scale which resulted into not rendering lensfalre at certain angles. The fix removes the looping through all cameras and uses the camera facing the light source to calculate occlusion. 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers


# Results 

![Screenshot from 2022-11-15 16-55-46](https://user-images.githubusercontent.com/28096692/202280539-e77c6641-116d-4601-8432-8fea8868bbe7.png)

sanjuksha committed:
Fix wide-angle camera lensflare occlusion (#3276)

* Fixing Z check in WideAngleCamera::Project3d

Signed-off-by: Sanjuksha Nirgude <sanjuksha@gmail.com>

Added a test case for wideangle camera lensflare

Signed-off-by: Aditya <aditya050995@gmail.com>

Add extra model in test world

The pre-existing model in the test world exhibits
the lens flare rendering bug without the fix in
this branch, so I added an extra model at a different
pose that does not exhibit the bug. It makes it
easier to confirm that the bug is fixed.

* Add extra cameras with occluded views
* Add another camera with box to its side that
  is incorrectly occluded

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

ahcorde commented on gazebo/rendering/WideAngleCamera.cc:
```diff
@@ -773,7 +773,8 @@ ignition::math::Vector3d WideAngleCamera::Project3d(
     pos.y /= pos.w;
     // check if point is visible
-    if (std::fabs(pos.x) <= 1 && std::fabs(pos.y) <= 1 && pos.z > 0)
+ if (std::fabs(pos.x) <= 1 && std::fabs(pos.y) <= 1 &&
+        pos.z > -std::fabs(pos.w))
```
indent properly

sanjuksha replied:
fixed

sanjuksha commented:
Added one more change for Z check based on this comment https://github.com/gazebosim/gazebo-classic/issues/3273#issuecomment-1307742254

scpeters commented:
cc @kjeppesen1 @mogumbo

iche033 commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -268,37 +271,27 @@ namespace gazebo
-            break;
+            ignition::math::Pose3d(
+            Conversions::ConvertIgn(cam->getDerivedPosition()),
+            Conversions::ConvertIgn(cam->getDerivedOrientation())));
```
can you indent these lines here?

kjeppesen1 commented:
This fix works for me :+1: I moved around a pair of wideangle cameras at various angles and positions in a world with a DEM, and I no longer see the lens flare disappearing when it isn't supposed to.

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -118,7 +118,10 @@ namespace gazebo
+
+
       {
+
```
Can you revert these whitespace changes (newlines)?

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -227,17 +232,15 @@ namespace gazebo
+            _wideAngleCam->AspectRatio());
         // aspect ratio should be 1.0 so VFOV should equal to HFOV
         this->dataPtr->wideAngleDummyCamera->SetHFOV(
             ignition::math::Angle(cam->getFOVy().valueRadians()));
```
I think this `cam` could be replaced with `_wideAngleCam` as well?

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -347,13 +340,14 @@ namespace gazebo
       }
       double s = static_cast<double>(rays - occluded) /
           static_cast<double>(rays);
+      // std::cout<<"Camera : "<< _cam->Name() <<" Occlusion Scale: "<< s << std::endl;
```
this line can be removed

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -268,37 +271,27 @@ namespace gazebo
-        for (auto cam : ogreEnvCameras)
-        {
           // project light world point to camera clip space.
           auto viewProj = cam->getProjectionMatrix() * cam->getViewMatrix();
```
replace `cam` with `_wideAngleCam` two times on this line

adityapande-1995 replied:
Hmm, ``_wideAngleCam`` doesn't have ``getProjectionMatrix()`` method here.

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -268,37 +271,27 @@ namespace gazebo
-                this->dataPtr->lightWorldPos);
-            break;
+            ignition::math::Pose3d(
+            Conversions::ConvertIgn(cam->getDerivedPosition()),
```
replace `cam` with `_wideAngleCam` on this and the following line

adityapande-1995 replied:
Same, ``_wideAngleCam`` doesn't have ``getDerivedPosition()`` here

scpeters commented on gazebo/rendering/LensFlare.cc:
```diff
@@ -212,8 +215,10 @@ namespace gazebo
+      std::vector<Ogre::Camera *> ogreEnvCameras = _wideAngleCam->OgreEnvCameras();
+      
+      // set dummy camera properties based on env cam
+      Ogre::Camera *cam = ogreEnvCameras[0];
```
I think the `cam` and `ogreEnvCameras` variables can be removed if all instances of `cam` are replaced by `_wideAngleCam`

iche033 replied:
actually just checking if `cam` should be replaced by `_wideAngleCam` in the calls below. I remember I had to use `cam` (one of the env cameras) because it has the right properties for doing ray casting. Can you do a quick check and print out the viewport size, FOV, and aspect ratio for `cam` vs `_wideAngleCam`? I think they could be different.

scpeters replied:
that's a good point

scpeters commented:
I've restarted the windows build

adityapande-1995 commented on gazebo/rendering/WideAngleCamera.cc:
```diff
@@ -773,7 +773,7 @@ ignition::math::Vector3d WideAngleCamera::Project3d(
     pos.x /= pos.w;
     pos.y /= pos.w;
     // check if point is visible
-    if (std::fabs(pos.x) <= 1 && std::fabs(pos.y) <= 1 && pos.z > 0)
+ if (std::fabs(pos.x) <= 1 && std::fabs(pos.y) <= 1 && pos.z > -std::fabs(pos.w))
```
Nit : Capitalize comment and fix indentation for the if statement: 
```
 // Check if point is visible
 if (std::fabs(pos.x) <= 1 && std::fabs(pos.y) <= 1 && pos.z > -std::fabs(pos.w))
```

adityapande-1995 replied:
You can add to this test : https://github.com/gazebosim/gazebo-classic/blob/5fef8ee173abef0eef54c816bc9c2a3089c6ffd1/test/integration/camera_sensor.cc#L1223

The included world already has a wide angle camera : https://github.com/gazebosim/gazebo-classic/blob/5fef8ee173abef0eef54c816bc9c2a3089c6ffd1/worlds/lensflare_plugin.world#L97

Could you add a small section that reflects the changes made here in this PR ?

adityapande-1995 replied:
Added a test : https://github.com/gazebosim/gazebo-classic/pull/3276/commits/3db832786de38fb6681b769a0de0855da3ca4db0

adityapande-1995 reviewed (changes requested):
It would be nice to add an example world or a test which has the camera in various poses, showing when the lensflare is supposed to appear and when it isn't.

scpeters commented:
I cleaned up some of the code changes in https://github.com/gazebosim/gazebo-classic/pull/3276/commits/5740ce52dc46363404be7d46f7fe986afe8d536b and https://github.com/gazebosim/gazebo-classic/pull/3276/commits/adf7d1f4b289af91b40e85240aa13e3b3671de6f and then made an adjustment to the new test in https://github.com/gazebosim/gazebo-classic/pull/3276/commits/c63047ca96959343f094a7c9ec012f150801f7ef to make it easier to confirm the presence of the bug without the fix in this branch

scpeters commented:
Thanks to some debugging help from @iche033, I added some additional cameras in https://github.com/gazebosim/gazebo-classic/pull/3276/commits/d33593d662e0c62cd85b1f2cb6abfeadba73877f and https://github.com/gazebosim/gazebo-classic/pull/3276/commits/c92f948e4116768ba788ba0bc79e1e4d2f5f6de4 to illustrate some problems with the occlusion checking in the test world added in this pull request

~~~
gazebo --verbose worlds/lensflare_wideangle_cam.world
~~~

### Failure to occlude in `wide_angle_cameras_occluded_high` camera image

There is a box directly in front of the camera, so it should occlude but doesn't

<img width="611" alt="Screen Shot 2023-01-31 at 3 26 07 PM" src="https://user-images.githubusercontent.com/3650755/215907415-cf53174d-4ff5-477d-b22b-aefa802d0c8f.png">

### Occluding when it shouldn't in `wide_angle_cameras_occluded_higher` camera image

There is a box to the right of the camera (partially visible), but it should not be occluding, but the lens flare is not there

<img width="614" alt="Screen Shot 2023-01-31 at 3 26 20 PM" src="https://user-images.githubusercontent.com/3650755/215907748-560c95d5-14d8-48fc-b6ad-d413a36ad116.png">

It seems to be checking for occlusions in the wrong direction?

mogumbo committed:
Improvements to WideAngleCamera LensFlare bug fix (#3296)

* Reverse extraneous z-up to y-up transform that was corrupting
  WideAngleCamera LensFlare results. Fixed small related bugs.

* Use fisheye cameras in lensflare_wideangle_cam.world to
  improve testing of lens flare.
